### PR TITLE
[MIG]base_import_module: Migrated to new API

### PR DESCRIPTION
**Pull request by msh-odoo**
Task: https://www.odoo.com/web?#id=12346&view_type=form&model=project.task&action=327
Pad: https://pad.odoo.com/p/r.qLBqdiKpoSjcFVeO

### CODE REVIEW

File: addons/base_import_module/__init__.py [hunk 1]
```diff
@@ -1,2 +1,22 @@
...
+#
+#    This program is distributed in the hope that it will be useful,
+#    but WITHOUT ANY WARRANTY; without even the implied warranty of
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU Affero General Public License for more details.
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
```
Comment (xmo-odoo) [inline addons/base_import_module/__init__.py]: Don't add that crap when it's not already there, especially in files as not-useful as an almost empty __init__.

File: addons/base_import_module/models/base_import_module.py [hunk 2]
```diff
@@ -1,49 +1,46 @@
...
+        module_obj = self.env['ir.module.module']
+        data = self.browse(self.ids[0])
         zip_data = base64.decodestring(data.module_file)
         fp = BytesIO()
         fp.write(zip_data)
-        res = module_obj.import_zipfile(cr, uid, fp, force=data.force, context=context)
-        self.write(cr, uid, ids, {'state': 'done', 'import_message': res[0]}, context=context)
-        context = dict(context, module_name=res[1])
+        res = module_obj.import_zipfile(fp, force=data.force)
+        self.browse(self.ids).write({'state': 'done', 'import_message': res[0]})
```
Comment (xmo-odoo) [inline addons/base_import_module/models/base_import_module.py]: Gets data from ids[0] but updates all the ids, is that really normal? Maybe it would make sense to just write on the already-browsed `data`?

File: addons/base_import_module/models/ir_module.py [hunk 3]
```diff
@@ -5,20 +6,20 @@
...
-class view(osv.osv):
+
+class View(models.Model):
     _inherit = "ir.module.module"
 
-    def import_module(self, cr, uid, module, path, force=False, context=None):
-        known_mods = self.browse(cr, uid, self.search(cr, uid, []))
+    def import_module(self, module, path, force=False):
+        known_mods = self.search([])
         known_mods_names = dict([(m.name, m) for m in known_mods])
```
Comment (xmo-odoo) [inline addons/base_import_module/models/ir_module.py]: use a dict comprehension, or at least remove the useless intermediate list comprehension

File: addons/base_import_module/__init__.py (hunk 1 above)
Comment (msh-odoo) [inline addons/base_import_module/__init__.py]: Isn't it required even in init file ? I added as per guidelines given by @jem-odoo

File: addons/base_import_module/models/base_import_module.py [hunk 4]
```diff
@@ -1,49 +1,46 @@
...
+    import_message = fields.Char()
+    force = fields.Boolean(string='Force init', help="Force init mode even if installed. (will update `noupdate='1'` records)")
 
-    def import_module(self, cr, uid, ids, context=None):
-        module_obj = self.pool.get('ir.module.module')
-        data = self.browse(cr, uid, ids[0] , context=context)
+    @api.multi
+    def import_module(self):
+        IrModule = self.env['ir.module.module']
+        data = self[0]
```
Comment (hmo-odoo) [inline addons/base_import_module/models/base_import_module.py]: what is this ? self[0]. I think, you should use self.ensure_one()

File: addons/base_import_module/models/base_import_module.py [hunk 5]
```diff
@@ -1,49 +1,46 @@
...
-            'res_id': ids[0],
+            'res_id': self.ids[0],
             'res_model': 'base.import.module',
             'type': 'ir.actions.act_window',
             'context': context,
         }
 
-    def action_module_open(self, cr, uid, ids, context):
+    @api.multi
+    def action_module_open(self):
```
Comment (hmo-odoo) [inline addons/base_import_module/models/base_import_module.py]: here need to put self.ensure_one

File: addons/base_import_module/models/ir_module.py [hunk 6]
```diff
@@ -5,21 +6,21 @@
...
-class view(osv.osv):
+
+class View(models.Model):
     _inherit = "ir.module.module"
 
-    def import_module(self, cr, uid, module, path, force=False, context=None):
-        known_mods = self.browse(cr, uid, self.search(cr, uid, []))
-        known_mods_names = dict([(m.name, m) for m in known_mods])
+    def import_module(self, module, path, force=False):
+        known_mods = self.search([])
```
Comment (hmo-odoo) [inline addons/base_import_module/models/ir_module.py]: why need to search all modules ? I think, we should optimize search query. first we should check unmet dependencies like

unmet_modules = self.search([('name', 'in', terp['depends']), ('state', '!=', 'installed')])
if unmet_modules:
   return error

File: addons/base_import_module/models/ir_module.py [hunk 7]
```diff
@@ -68,15 +69,15 @@ def import_module(self, cr, uid, module, path, force=False, context=None):
                         type='binary',
                         datas=data,
                     )
-                    att_id = ir_attach.search(cr, uid, [('url', '=', url_path), ('type', '=', 'binary'), ('res_model', '=', 'ir.ui.view')], context=context)
-                    if att_id:
-                        ir_attach.write(cr, uid, att_id, values, context=context)
+                    att = IrAttachment.search([('url', '=', url_path), ('type', '=', 'binary'), ('res_model', '=', 'ir.ui.view')])
```
Comment (hmo-odoo) [inline addons/base_import_module/models/ir_module.py]: rename variable 'att' with 'attachment'

File: addons/base_import_module/models/ir_module.py (hunk 6 above)
Comment (msh-odoo) [inline addons/base_import_module/models/ir_module.py]: May be you are partially right, known_mods are further used to get module which going to import.

File: addons/base_import_module/__openerp__.py [hunk 8]
```diff
@@ -11,7 +13,7 @@
     'depends': ['web'],
     'installable': True,
     'auto_install': False,
-    'data': ['views/base_import_module.xml'],
+    'data': ['views/base_import_module_view.xml'],
     'qweb': [],
     'test': [],
```
Comment (WHPODOO) [inline addons/base_import_module/__openerp__.py]: If not use then remove it : 
```
'qweb': [],
'test': [],
```

File: addons/base_import_module/models/base_import_module.py [hunk 9]
```diff
@@ -1,49 +1,46 @@
...
+            'res_id': self.ids,
             'res_model': 'base.import.module',
             'type': 'ir.actions.act_window',
             'context': context,
         }
 
-    def action_module_open(self, cr, uid, ids, context):
+    @api.multi
+    def action_module_open(self):
+        self.ensure_one()
```
Comment (WHPODOO) [inline addons/base_import_module/models/base_import_module.py]: Its already put here :last_quarter_moon:

File: addons/base_import_module/tests/test_module/__openerp__.py [hunk 10]
```diff
@@ -9,6 +9,7 @@
     'description': """
         Test
         """,
+    'author': 'Odoo SA',
```
Comment (WHPODOO) [inline addons/base_import_module/tests/test_module/__openerp__.py]: No need to write : 
```
'author': 'Odoo SA',
```

File: addons/base_import_module/controllers/main.py [hunk 11]
```diff
@@ -15,21 +15,17 @@ def wrap(*args, **kw):
 class ImportModule(Controller):
 
     def check_user(self, uid=None):
-        if uid is None:
-            uid = request.uid
-        is_admin = request.registry['res.users']._is_admin(request.cr, uid, [uid])
-        if not is_admin:
+        if not request.env.user._is_admin():
```
Comment (msh-odoo) [inline addons/base_import_module/controllers/main.py]: why checking with environment's user, why not checking with uid passed in method parameter ? what if I run deploy command with demo user ?
Try with this scenario: run your server, open new terminal and run following command
```./odoo.py deploy --db=your_loaded_db(with base_import_module installed) --login=demo --password=demo path='module directory path'```

File: addons/base_import_module/controllers/main.py [hunk 12]
```diff
@@ -15,21 +15,17 @@ def wrap(*args, **kw):
...
     @webservice
     def login(self, login, password, db=None):
         if db and db != request.db:
             raise Exception("Could not select database '%s'" % db)
-        uid = request.session.authenticate(request.db, login, password)
-        if not uid:
+        if not request.session.authenticate(request.db, login, password):
             return Response(response="Wrong login/password", status=401)
-        self.check_user(uid)
+        self.check_user()
```
Comment (msh-odoo) [inline addons/base_import_module/controllers/main.py]: call check_user method with uid returned from authenticate method

File: addons/base_import_module/controllers/main.py [hunk 13]
```diff
@@ -15,21 +15,17 @@ def wrap(*args, **kw):
 class ImportModule(Controller):
 
     def check_user(self, uid=None):
-        if uid is None:
-            uid = request.uid
```
Comment (msh-odoo) [inline addons/base_import_module/controllers/main.py]: May be we should keep this code, if uid is None then we will consider uid = request.env.user.id

File: addons/base_import_module/models/ir_module.py [hunk 14]
```diff
@@ -5,41 +6,33 @@
 from os.path import join as opj
 
 import openerp
-from openerp import fields
-from openerp.osv import osv
+from openerp import models, _
```
Comment (msh-odoo) [inline addons/base_import_module/models/ir_module.py]: Use from odoo .... (applies to all file)

File: addons/base_import_module/controllers/main.py [hunk 15]
```diff
@@ -16,12 +16,12 @@ class ImportModule(Controller):
 
     def check_user(self, uid=None):
         if uid is None:
-            uid = request.uid
-        is_admin = request.registry['res.users']._is_admin(request.cr, uid, [uid])
+            uid = request.env.user.id
+        is_admin = request.env['res.users'].browse(int(uid))._is_admin()
```
Comment (msh-odoo) [inline addons/base_import_module/controllers/main.py]: Do we need to cast explicitly ?

File: addons/base_import_module/controllers/main.py [hunk 16]
```diff
@@ -16,12 +16,12 @@ class ImportModule(Controller):
 
     def check_user(self, uid=None):
         if uid is None:
-            uid = request.uid
-        is_admin = request.registry['res.users']._is_admin(request.cr, uid, [uid])
+            uid = request.env.user.id
+        is_admin = request.env['res.users'].browse(uid)._is_admin()
```
Comment (hmo-odoo) [inline addons/base_import_module/controllers/main.py]: no need browse again. use self.env.user

File: addons/base_import_module/controllers/main.py [hunk 17]
```diff
@@ -16,12 +16,12 @@ class ImportModule(Controller):
...
         if not is_admin:
-            raise openerp.exceptions.AccessError("Only administrators can upload a module")
+            raise odoo.exceptions.AccessError("Only administrators can upload a module")
```
Comment (hmo-odoo) [inline addons/base_import_module/controllers/main.py]: error message should be translated

File: addons/base_import_module/controllers/main.py [hunk 18]
```diff
@@ -16,12 +16,12 @@ class ImportModule(Controller):
...
 
-    @route('/base_import_module/login', type='http', auth='none', methods=['POST'])
+    @route('/base_import_module/login', type='http', auth='none', methods=['POST'], csrf=False)
```
Comment (hmo-odoo) [inline addons/base_import_module/controllers/main.py]: can you make separate PR to handle csrf security ?

File: addons/base_import_module/controllers/main.py [hunk 19]
```diff
@@ -37,7 +37,7 @@ def login(self, login, password, db=None):
             'X-CSRF-TOKEN': request.csrf_token(),
         })
 
-    @route('/base_import_module/upload', type='http', auth='user', methods=['POST'])
+    @route('/base_import_module/upload', type='http', auth='user', methods=['POST'], csrf=False)
```
Comment (tde-banana-odoo) [inline addons/base_import_module/controllers/main.py]: No

Comment (tivisse) [inline addons/base_import_module/controllers/main.py]: I'm not sure that removing a security layer is in the scope of the migration. Or is it to avoid something ? I hope not. I would be very angry. And when I'm angry I want to break things.

File: addons/base_import_module/models/ir_module.py [hunk 20]
```diff
@@ -1,45 +1,45 @@
...
+class IrModule(models.Model):
     _inherit = "ir.module.module"
 
     imported = fields.Boolean(string="Imported Module", default=False)
 
-    def import_module(self, cr, uid, module, path, force=False, context=None):
-        known_mods = self.browse(cr, uid, self.search(cr, uid, []))
-        known_mods_names = dict([(m.name, m) for m in known_mods])
+    def import_module(self, module, path, force=False):
+        known_mods = self.search(cr, uid, [])
```
Comment (tivisse) [inline addons/base_import_module/models/ir_module.py]: That will not work: remaining cr, uid

File: addons/base_import_module/models/ir_module.py [hunk 21]
```diff
@@ -71,17 +71,17 @@ def import_module(self, cr, uid, module, path, force=False, context=None):
...
+                    attachment = IrAttachment.search([('url', '=', url_path), ('type', '=', 'binary'), ('res_model', '=', 'ir.ui.view')])
+                    if attachment:
+                        attachment.write(values)
                     else:
-                        ir_attach.create(cr, uid, values, context=context)
+                        IrAttachment.create(values)
 
         return True
 
-    def import_zipfile(self, cr, uid, module_file, force=False, context=None):
```
Comment (tivisse) [inline addons/base_import_module/models/ir_module.py]: api.model
